ctrd/image_commit.go

```diff
@@ -0,0 +1,326 @@
+	containerdUncompressed = "containerd.io/uncompressed"
+	manifestType           = images.MediaTypeDockerSchema2Manifest
+	configType             = images.MediaTypeDockerSchema2Config
+	layerType              = images.MediaTypeDockerSchema2LayerGzip
+)
+
+// CommitConfig defines options for committing a container image
+type CommitConfig struct {
+
+	// author
+	Author string
+
+	// comment
+	Comment string
+
+	// container config
+	ContainerConfig *types.ContainerConfig
+
+	// container ID
+	ContainerID string
+
+	// parent reference
+	ParentReference string
+
+	// reference
+	Reference string
+
+	// repository
+	Repository string
+
+	// containerd format image
+	CImage containerd.Image
+
+	// image-spec format image
+	Image ocispec.Image
+}
+
+// Commit commits an image from a container.
+func (c *Client) Commit(ctx context.Context, config *CommitConfig) (imageID string, err0 error) {
+	// get a containerd client
+	wrapperCli, err := c.Get(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to get a containerd grpc client: %v", err)
+	}
+	client := wrapperCli.client
+
+	var (
+		sn     = client.SnapshotService(defaultSnapshotterName)
+		cs     = client.ContentStore()
+		differ = client.DiffService()
+	)
+
+	// export new layer
+	snapshot, err := c.GetSnapshot(ctx, config.ContainerID)
+	if err != nil {
+		return "", fmt.Errorf("failed to get snapshot: %s", err)
+	}
+	layer, diffIDStr, err := exportLayer(ctx, snapshot.Name, sn, cs, differ)
+	if err != nil {
+		return "", err
+	}
+
+	// create child image
+	diffIDDigest, err := digest.Parse(diffIDStr)
+	if err != nil {
+		return "", err
+	}
+
+	childImg, err := newChildImage(ctx, config, diffIDDigest)
+	if err != nil {
+		return "", err
+	}
+
+	// create new snapshot for new layer
+	snapshotKey := identity.ChainID(childImg.RootFS.DiffIDs).String()
+	if err = newSnapshot(ctx, config.Image, sn, differ, layer, snapshotKey, diffIDStr); err != nil {
+		return "", err
+	}
+	defer func() {
+		if err0 != nil {
+			logrus.Warnf("remove snapshot %s cause commit image failed", snapshotKey)
+			client.SnapshotService(defaultSnapshotterName).Remove(ctx, snapshotKey)
+		}
+	}()
+
+	imgJSON, err := json.Marshal(childImg)
+	if err != nil {
+		return "", err
+	}
+
+	// new config descriptor
+	configDesc := ocispec.Descriptor{
+		MediaType: configType,
+		Digest:    digest.FromBytes(imgJSON),
+		Size:      int64(len(imgJSON)),
+	}
+
+	// get parent image layer descriptor
+	pmfst, err := images.Manifest(ctx, cs, config.CImage.Target(), platforms.Default())
+	if err != nil {
+		return "", err
+	}
+
+	// new layer descriptor
+	layers := append(pmfst.Layers, layer)
+	labels := map[string]string{
+		"containerd.io/gc.ref.content.0": configDesc.Digest.String(),
+	}
+	for i, l := range layers {
+		labels[fmt.Sprintf("containerd.io/gc.ref.content.%d", i+1)] = l.Digest.String()
+	}
+
+	// new manifest descriptor
+	mfst := struct {
```

Comment:
why not use https://github.com/opencontainers/image-spec/blob/master/specs-go/v1/manifest.go#L20 here?